controllers/kserve_raw_inferenceservice_controller_test.go

```diff
@@ -0,0 +1,185 @@
+			inferenceService := createInferenceService(testNs, KserveOvmsInferenceServiceName, KserveInferenceServicePath1)
+			inferenceService.Spec.Predictor.ServiceAccountName = serviceAccountName
+			if err := cli.Create(ctx, inferenceService); err != nil && !apierrs.IsAlreadyExists(err) {
+				Expect(err).NotTo(HaveOccurred())
+			}
+
+			crb := &rbacv1.ClusterRoleBinding{}
+			Eventually(func() error {
+				key := types.NamespacedName{Name: inferenceService.Namespace + "-" + serviceAccountName + "-auth-delegator",
+					Namespace: inferenceService.Namespace}
+				return cli.Get(ctx, key, crb)
+			}, timeout, interval).ShouldNot(HaveOccurred())
+		})
+		It("it should create a route if isvc has the label to expose route", func() {
+			_ = createServingRuntime(testNs, KserveServingRuntimePath1)
+			inferenceService := createInferenceService(testNs, KserveOvmsInferenceServiceName, KserveInferenceServicePath1)
+			inferenceService.Labels = map[string]string{}
+			inferenceService.Labels[constants.KserveNetworkVisibility] = constants.LabelEnableKserveRawRoute
+			if err := cli.Create(ctx, inferenceService); err != nil && !apierrs.IsAlreadyExists(err) {
+				Expect(err).NotTo(HaveOccurred())
+			}
+			isvcService := &corev1.Service{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      KserveOvmsInferenceServiceName + "-predictor",
+					Namespace: inferenceService.Namespace,
+					Annotations: map[string]string{
+						"openshift.io/display-name":        KserveOvmsInferenceServiceName,
+						"serving.kserve.io/deploymentMode": "RawDeployment",
+					},
+					Labels: map[string]string{
+						"app":                                "isvc." + KserveOvmsInferenceServiceName + "-predictor",
+						"component":                          "predictor",
+						"serving.kserve.io/inferenceservice": KserveOvmsInferenceServiceName,
+					},
+				},
+				Spec: corev1.ServiceSpec{
+					ClusterIP:  "None",
+					IPFamilies: []corev1.IPFamily{"IPv4"},
+					Ports: []corev1.ServicePort{
+						{
+							Name:       "https",
+							Protocol:   corev1.ProtocolTCP,
+							Port:       8888,
+							TargetPort: intstr.FromString("https"),
+						},
+					},
+					ClusterIPs: []string{"None"},
+					Selector: map[string]string{
+						"app": "isvc." + KserveOvmsInferenceServiceName + "-predictor",
+					},
+				},
+			}
+			if err := cli.Create(ctx, isvcService); err != nil && !apierrs.IsAlreadyExists(err) {
+				Expect(err).NotTo(HaveOccurred())
+			}
+			service := &corev1.Service{}
+			Eventually(func() error {
+				key := types.NamespacedName{Name: isvcService.Name, Namespace: isvcService.Namespace}
+				return cli.Get(ctx, key, service)
+			}, timeout, interval).Should(Succeed())
+			route := &routev1.Route{}
+			Eventually(func() error {
+				key := types.NamespacedName{Name: inferenceService.Name, Namespace: inferenceService.Namespace}
+				return cli.Get(ctx, key, route)
+			}, timeout, interval).ShouldNot(HaveOccurred())
```

Comment:
About your comment in Slack, in CI I'm seeing this check failing.

In CI, I do see in the logs of the controller `Failed to fetch service for InferenceService`. However, that log message does not necessarily come because of an error. I see the code:

```go
if err != nil || len(serviceList.Items) == 0 {
		log.Error(err, "Failed to fetch service for InferenceService", "InferenceService", isvc.Name)
		return nil, err
}
```

Most likely, `err` is nil, but the query didn't return any items. Effectively, it returns `(nil, nil)` So, the reconcile continues normally as if no route is wanted and without dropping any error that would retry the reconcile. This would be leading to a failing test here.

I see this just as a race condition where the reconcile finishes before this unit is able to create the Service.

Reply:
I moved the service creation before the isvc to avoid the race condition.